### ISSUE TITLE
Texture3D direct data access

### ISSUE THREAD

**Issue by thmasn** [godotengine/godot-proposals#9583]
### Describe the project you are working on

i am creating a voxel-game which heavily relies on [Texture3D ](https://docs.godotengine.org/en/stable/classes/class_texture3d.html#class-texture3d-private-method-get-data) as a structure for voxel data.
creating and modifying Textures3D happens a lot in the code.

also, non-voxel games use Texture3D for fog, windmaps, density maps, etc.

### Describe the problem or limitation you are having in your project

since there is no direct way of accessing Texture3D data, modifying a Texture3D requires a lot of GDscript.
p.e. to change a single pixel, you need to use the following code:
```
func setBlock(coor:Vector3i, block:int):
	var data := tex3d.get_data() #get the Texture3D data
	var img := data[coor.z] # get the correct image slice for this z-coordinate
	var imgData := img.get_data() #get the image data
	imgData[coor.x + coor.y*chunk.get_width()] = block #set the pixel data in the slice
	data[coor.z].set_data(tex3d.get_width(), tex3d.get_height(), false, Image.FORMAT_L8, imgData) #replace the slice data
	tex3d.update(data) #update the texture3D
```
replacing all data on a texture3d requires the following code:
```
func fromPackedByteArray(byteArray:PackedByteArray, size:Vector3i):
	var layers : Array[Image] # create our image array
	var zSectionSize := size.x*size.y # the size we will slice the bytearray into
	for z in size.z: # fore each z-slice
		var image := Image.create(8,8,false,ImageFormat) # create a new image
		image.set_data(size.x, size.y, false, ImageFormat,byteArray.slice(z*zSectionSize,(z+1)*zSectionSize)) #get the data for this slice
		layers.append(image) # add our image to the array
	tex3d= ImageTexture3D.new() # create a new texture3D
	tex3d.create(Image.FORMAT_L8, size.x, size.y, size.x, false, layers) #fill texture3D with data
```

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

additional functions to modify Texture3D data directly without going through the extra layer of the image array

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

similar functions to which are already present on [Image](https://docs.godotengine.org/en/stable/classes/class_image.html#class-image):
`void set_pixel( int x, int y, int z, Color color )` # set one pixel in a texture3D
`void set_pixelv( Vector2i point, Color color )` # set one pixel in a texture3D
`PackedByteArray get_data( )` # get the complete data of the texture3D as a bytearray
`set_data( int width, int height, int depth, bool use_mipmaps, Format format, PackedByteArray data )` # replace all data with a bytearray

### If this enhancement will not be used often, can it be worked around with a few lines of script?

it can currently be done with gdscript, but a direct access would reduce the amount of code by a factor of 5.
hopfully, direct access would also improve performance, as creating the images could potentially be skipped?

### Is there a reason why this should be core and not an add-on in the asset library?

this is basic functionality for texture3d, which is present in other game engines.

**Comment by clayjohn** [comment]
Remember that Images and Textures are two very distinct concepts. An image is a CPU representation of a bunch of pixels. Which is why it has things like `set_pixel()`. A Texture is a GPU resource that is used for rendering. It is also a collection of pixels, but it resides fully on the GPU for performance reasons. The workflow in Godot is for CPU manipulation to happen with Images, and then you update the texture using the Image. This workflow is pretty important for performance reasons. 

What we are missing right now is partial texture updates, which would allow you to update only a sub-region of a texture using an image

**Comment by thmasn** [comment]
true, potentially it would mean introducing an Image3D as a counterpart?
then, the Texture3D.get_data() could return an Image3D, which can have all the set_pixel functionality, and then be updated back into the Texture3D?
the current inconvenience is going over an Image-Array, which introduces all the slicing and updating happening in the example code above.

**Comment by clayjohn** [comment]
> true, potentially it would mean introducing an Image3D as a counterpart? then, the Texture3D.get_data() could return an Image3D, which can have all the set_pixel functionality, and then be updated back into the Texture3D? the current inconvenience is going over an Image-Array, which introduces all the slicing and updating happening in the example code above.

Image3D could work. Or maybe even just a function to get a slice (which would be an Image)
